consensus/src/lib.rs

```diff
@@ -166,30 +180,53 @@ impl<N: GossipNetwork, S: Storage> Consensus<N, S> {
         Ok(())
     }
 
-    pub async fn read(&self) -> Result<ConsensusState, Error> {
-        Ok(self.state.consensus_state.clone())
-    }
+    // Todo: Read public state from the vetomint FSM.
+    // pub async fn read_consensus_state(&self) -> Result<ConsensusState, Error> {
+    //     Ok(self.state.vetomint.state)
+    // }
 
-    pub async fn veto_block(
+    pub async fn set_proposal_candidate(
         &mut self,
-        _network_config: NetworkConfig,
-        _known_peers: &[Peer],
-        _block_hash: Hash256,
-    ) -> Result<(), Error> {
-        unimplemented!()
+        network_config: &NetworkConfig,
+        known_peers: &[Peer],
+        block_hash: Hash256,
+        timestamp: Timestamp,
+    ) -> Result<Vec<ProgressResult>, Error> {
+        self.abort_if_finalized()?;
+        let block_index = self.get_block_index(&block_hash)?;
+        let consensus_event = ConsensusEvent::BlockCandidateUpdated {
+            proposal: block_index,
+        };
+        let responses = self.state.vetomint.progress(consensus_event, timestamp);
+        let result = self
+            .process_multiple_responses(responses, network_config, known_peers, timestamp)
+            .await?;
+        Ok(result)
     }
 
-    pub async fn set_proposal(&mut self, _block_hash: Hash256) -> Result<(), Error> {
-        unimplemented!()
+    pub async fn veto_block(&mut self, block_hash: Hash256) -> Result<(), Error> {
+        self.abort_if_finalized()?;
+        self.state.vetoed_block_hashes.push(block_hash);
+        Ok(())
     }
 
+    #[allow(unreachable_code, clippy::diverging_sub_expression)]
```

Comment:
What is this attribution for? Can you tell me the situation?

Reply:
fixed